### DISCUSSION TITLE
Powering a bunch of raspberry pis and hifiberries

### DISCUSSION THREAD

**Discussion by cxrodgers** [wehr-lab/autopilot#126] [General]
So, the hifiberry spec sheet recommends an 18V separate power supply. It also says you can actually power the raspberry pi itself through the hifiberry, just using that 18V supply.

Instead, what I've been doing is just using the regular CanaKit 5V 3A brick to power the raspberry pi, and this seems to be plenty to run the pi and the hifiberry. I wonder if this is because I'm just running tweeters with the hifiberry, which probably take a lot less power than the house-party-style speakers that the hifiberry designers had in mind.

Is there any reason to think that 18V supply is actually necessary? Am I somehow degrading the quality of my noise bursts by not using it?

On a related note, as my collection of raspberry pis grows, these CanaKit bricks are taking up all my outlets. I was thinking about getting a USB power supply hub like this one (https://www.amazon.com/Anker-Charger-PowerPort-iPhone-Galaxy/dp/B00P936188). It's supposed to do 2.4A per port. Then I can connect each raspberry pi to the hub with a nice flexible USB-C cable, instead of that bulky CanaKit barrel jack. Does this sound reasonable? 

Alternatively, I could take one high-current 18V supply and split its output to power many hifiberries with red/black wire pairs, but I would worry that I'd have a short at one of the screw terminals and it would take down the whole network.

**Comment by sneakers-the-rat** [comment]
I was just following the spec! I did have some low power issues with 2.5A power supplies when using Pi3Bs, but if you're not experiencing that then by all means do use them! We can add this to the hifiberry page as well :). it's definitely the case that i need to scale the amplitude of all my sounds to like 1% max amplitude because otherwise it would be totally deafening (100% need to build in sound calibration 🙃)

The biggest problem with not having a power supply with plenty of headroom is that it can cause unpredictable errors in the case of a spike in power draw as the CPU undervolts. Since that might happen at critical moments (eg. when a sound is played) I have opted to just not think about it by using big fat power supplies, but that's from a place of paranoia not reason :p.  Since it's a hardware glitch it's really hard to know what to expect, mostly it'll be slowdowns/crashing, but i have also had nonsensical data being sent in messages. Undervolt logging is [included in the raspi kernel](https://github.com/raspberrypi/linux/pull/2397) i believe, so you should be able to see if you're getting undervolts from the kernel log -- i believe that's `journalctl -k`. 

I'm not a very sophisticated electronics person, but yes i would also be worried about sharing a power supply across pis for that reason, seems like if one pi had a transient increase in current draw it could starve the others? but maybe not, sounds like an empirical question.

Relatedly, and as a tangent, I also had problems from transient undervolting from ~2.5A supplies when there would be a clock frequency switch in the raspi cpu. The clock switches can also cause audio artifacts, so in the `performance` script I set the cpu governor to always be on 'performance' mode. Just to say another issue i've found with using power supplies that can supply stable amperage but suffer with spikes for the sake of collective memory: https://github.com/wehr-lab/autopilot/blob/dd3e3aae8d96c6f5665069842e5fc74e2a84e009/autopilot/setup/scripts.py#L75

**Reply by cxrodgers** [reply]
That's a really great point which I had not considered. I briefly googled typical current draw of a 4B and people say it's like <1A, but of course the worst case scenario is the transient spike leading to undervolting. I'll check my kernel logs for any warning messages. Setting the CPU to have a constant clock frequency seems like generally a good idea too.

In any case I ordered the USB-C power hub, so I'll see how it works.

**Reply by cxrodgers** [reply]
So far so good with the USB-C power hub. I've been checking for undervoltage with `vcgencmd get_throttled` (which returns 0x0 unless there's been an event since the last reboot) and inspecting `journalctl -k`. 

Coincidentally, there was a MASSIVE undervoltage event in NYC over the weekend. At 0:07 in this video you can see the entire city go dim for an instant. https://www.reddit.com/r/nyc/comments/pee6bq/video_of_con_edison_explosion_from_earthcam_4k/

Unfortunately, at my workplace this was enough to completely take out the power for a few minutes, instead of just undervolting, so my Pis rebooted and I wasn't able to check how this affected the logs :)
